pkg/operator/ceph/cluster/osd/osd.go

```diff
@@ -126,68 +126,94 @@ type OSDInfo struct {
-				Name: hostname,
+	if len(c.ValidStorage.VolumeSources) == 0 {
+		if c.DesiredStorage.UseAllNodes {
+			// Get the list of all nodes in the cluster. The placement settings will be applied below.
+			hostnameMap, err := k8sutil.GetNodeHostNames(c.context.Clientset)
+			if err != nil {
+				logger.Warningf("failed to get node hostnames: %v", err)
+				return err
 			}
-			c.DesiredStorage.Nodes = append(c.DesiredStorage.Nodes, storageNode)
+			c.DesiredStorage.Nodes = nil
+			for _, hostname := range hostnameMap {
+				storageNode := rookalpha.Node{
+					Name: hostname,
+				}
+				c.DesiredStorage.Nodes = append(c.DesiredStorage.Nodes, storageNode)
+			}
+			logger.Debugf("storage nodes: %+v", c.DesiredStorage.Nodes)
 		}
-		logger.Debugf("storage nodes: %+v", c.DesiredStorage.Nodes)
-	}
-	// generally speaking, this finds nodes which are capable of running new osds
-	validNodes := k8sutil.GetValidNodes(c.DesiredStorage, c.context.Clientset, c.placement)
-
-	// no valid node is ready to run an osd
-	if len(validNodes) == 0 {
-		logger.Warningf("no valid node available to run an osd in namespace %s. "+
-			"Rook will not create any new OSD nodes and will skip checking for removed nodes since "+
-			"removing all OSD nodes without destroying the Rook cluster is unlikely to be intentional", c.Namespace)
-		return nil
+		// generally speaking, this finds nodes which are capable of running new osds
+		validNodes := k8sutil.GetValidNodes(c.DesiredStorage, c.context.Clientset, c.placement)
+
+		// no valid node is ready to run an osd
+		if len(validNodes) == 0 && len(c.ValidStorage.VolumeSources) == 0 {
+			logger.Warningf("no valid node available to run an osd in namespace %s. "+
+				"Rook will not create any new OSD nodes and will skip checking for removed nodes since "+
+				"removing all OSD nodes without destroying the Rook cluster is unlikely to be intentional", c.Namespace)
+			return nil
+		}
+		logger.Infof("%d of the %d storage nodes are valid", len(validNodes), len(c.DesiredStorage.Nodes))
+		c.ValidStorage = *c.DesiredStorage.DeepCopy()
+		c.ValidStorage.Nodes = validNodes
+	} else {
+		c.ValidStorage.Nodes = nil
 	}
-	logger.Infof("%d of the %d storage nodes are valid", len(validNodes), len(c.DesiredStorage.Nodes))
-	c.ValidStorage = *c.DesiredStorage.DeepCopy()
-	c.ValidStorage.Nodes = validNodes
 
 	// start the jobs to provision the OSD devices and directories
-	config := newProvisionConfig()
+
 	logger.Infof("start provisioning the osds on nodes, if needed")
 	c.startProvisioning(config)
-
 	// start the OSD pods, waiting for the provisioning to be completed
 	logger.Infof("start osds after provisioning is completed, if needed")
 	c.completeProvision(config)
 
 	// handle the removed nodes and rebalance the PGs
-	logger.Infof("checking if any nodes were removed")
-	c.handleRemovedNodes(config)
+	if len(c.ValidStorage.VolumeSources) == 0 {
+		logger.Infof("checking if any nodes were removed")
+		c.handleRemovedNodes(config)
```

Comment:
`handleRemoveNodes()` should be a no-op for device sets because it relies on the `nodes` element in the CR, which device sets ignore.

Reply:
cc @ashishranjan738